contributors/design-proposals/deploy.md

```diff
@@ -118,29 +124,35 @@ See issue [#17164](https://github.com/kubernetes/kubernetes/issues/17164).
-deployment history, and then we do a rollout with the same PodTemplate as version 1, then version 1 is lost and becomes version 4 after the rollout.
+Note that a Deployment's ReplicaSets use PodTemplate hashes (i.e. the hash of `.spec.template`) to distinguish from each others. When doing rollout or rollback, a Deployment reuses existing ReplicaSet if it has the same PodTemplate, and its `kubernetes.io/change-cause` and `deployment.kubernetes.io/revision` annotations will be updated by the new rollout. All previous of versions of this ReplicaSet will be kept in the annotation `deployment.kubernetes.io/revision-history`. For example, if we had 3 ReplicaSets in
+Deployment history, and then we do a rollout with the same PodTemplate as version 1, then version 1 is lost and becomes version 4 after the rollout, and the ReplicaSet that once represented version 1 will then have an annotation `deployment.kubernetes.io/revision-history=1`.
 
-To make deployment versions more meaningful and readable for the users, we can add more annotations in the future. For example, we can add the following flags to `kubectl` for the users to describe and record their current rollout:
+To make Deployment versions more meaningful and readable for the users, we can add more annotations in the future. For example, we can add the following flags to `kubectl` for the users to describe and record their current rollout:
 - `--description`: adds `description` annotation to an object when it's created to describe the object.
 - `--note`: adds `note` annotation to an object when it's updated to record the change.
 - `--commit`: adds `commit` annotation to an object with the commit id.
 
 ### Pause Deployments
 
-Users sometimes need to temporarily disable a deployment. See issue [#14516](https://github.com/kubernetes/kubernetes/issues/14516).
+Users sometimes need to temporarily disable a Deployment. See issue [#14516](https://github.com/kubernetes/kubernetes/issues/14516).
+
+For more details, see [pausing and resuming a
+Deployment](https://kubernetes.io/docs/user-guide/deployments/#pausing-and-resuming-a-deployment).
 
 ### Perm-failed Deployments
 
-The deployment could be marked as "permanently failed" for a given spec hash so that the system won't continue thrashing on a doomed deployment. The users can retry a failed deployment with `kubectl rollout retry`. See issue [#14519](https://github.com/kubernetes/kubernetes/issues/14519).
+The Deployment could be marked as "permanently failed" for a given spec hash so that the system won't continue thrashing on a doomed Deployment. See issue [#14519](https://github.com/kubernetes/kubernetes/issues/14519).
```

Comment:
I think we are misusing the term "perma-failed". Once we identify a permanent failure, we should be able to abort the rollout but in reality I don't think we can identify those reliably, that's why we still haven't ended up doing something about it (imagePullBackOff, crashLoopBackOff, insufficient quota, or permissions, all can be transient errors). ProgressDeadlineSeconds merely bubbles a Progressing condition with reason=ProgressDeadlineExceeded in the Deployment status. This can be useful for external processes now and in the future I expect we will automatically rollback as soon as we observe the condition (users now could easily implememnt autoRollback by writting an external controller that would populate d.Spec.Rollback as soon as a ProgressDeadlineExceeded is observed).

TL; DR
Let's stop using "permanently". Switch the title as well to "Failed" or "Stuck".

Reply:
Updated, PTAL